core/include/drivers/clk.h

```diff
@@ -0,0 +1,193 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#ifndef CLK_H
+#define CLK_H
+
+#include <kernel/refcount.h>
+#include <kernel/spinlock.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+
+/* Flags for clock */
+#define CLK_SET_RATE_GATE	BIT(0) /* must be gated across rate change */
+#define CLK_SET_PARENT_GATE	BIT(1) /* must be gated across re-parent */
+
+/**
+ * struct clk - Clock structure
+ *
+ * @name: Clock name
+ * @priv: Private data for the clock provider
+ * @ops: Clock operations
+ * @parents: Array of parents of the clock
+ * @parent: Current parent
+ * @num_parents: Number of parents
+ * @rate: Current clock rate (cached after init or rate change)
+ * @flags: Specific clock flags
+ * @lock: Lock for atomic clock operations
+ * @enabled_count: Clock state (cached)
+ */
+struct clk {
+	const char *name;
+	void *priv;
+	const struct clk_ops *ops;
+	struct clk **parents;
```

Comment:
We could save a pointer by putting this at the end of this struct instead as:
``` C
struct clk *parents[];
```

Reply:
Yes indeed ! Didn't thought about this one !